Added the feature to display the external link in the warning dialog box

shankarpriyank opened the pull request:
<!--
- Add the issue number here.

- If you haven't solved the issue completely use "Linked issue #{issue_number}.
- After solving the issue completely change it to "Fixes #{issue_number}.
-->
Fixes #2966
Fixes #2965

<!-- Add here what changes were made in this issue and if possible provide links. -->
Made changes to files related to Dialog Boxes, to show the URL of external link in the warning dialog box 


<!-- If possible, please add relevant screenshots / GIFs -->

**Screenshots** 
Light mode       |  Dark mode |
:-------------------------:|:-------------------------:
![image](https://user-images.githubusercontent.com/34593983/216536338-2accf48d-bfc0-4b41-bf3a-4ddd72be066e.jpg)   |     ![image](https://user-images.githubusercontent.com/34593983/216536370-338d173b-69f3-490c-a26c-afa3dd6ffeaf.jpg)

kelson42 commented:
Few remarks:
* I would remove the "Link - "
* Buttons should be better aligned, "Yes" and "no" aligned left and on a first ligne, the "Don't ask anymore" left aligned on a second lign

shankarpriyank commented:
Okay, willl remove "Link-"
Sorry but I cannot comprehend what do you mean by
  > "Yes" and "no" aligned left and on a first ligne, the "Don't ask anymore" left aligned on a second lign
 
 I guess all there buttons are on three different lines, also I did not  did not change anything related to the buttons.
   
   This is how the dialog box looks on the develop branch 
   
![image](https://user-images.githubusercontent.com/75121767/211146448-fff61411-08e4-4f5a-a183-266b70d0c158.png)

kelson42 commented:
@shankarpriyank Don't remove the link, remove "Link -" label. We know this is a lonk, not needed to explain it.

shankarpriyank commented:
Hey @kelson42  I have removed the label, and also fixed the failing tests.
Thought the lint checks will still fail, the error I am getting is this
` This expression contains a magic number. Consider defining it to a well named constant. [MagicNumber]
`
This is thrown at a line where I am assigning a value to a variable, I am unable to understand how to fix this.

kelson42 commented:
@shankarpriyank Thank you, @gouri-panda will review your work. Give him a few days.

shankarpriyank commented:
Updated Screenshot
![image](https://user-images.githubusercontent.com/75121767/211162708-4c1dd42b-a8d6-40ae-89eb-6fc5fb5274b1.png)

gouri-panda commented:
>This expression contains a magic number. Consider defining it to a well named constant. [MagicNumber]
This is thrown at a line where I am assigning a value to a variable, I am unable to understand how to fix this.

@shankarpriyank You can suppress this by annotating method supress warnings or make a constant value.

shankarpriyank commented:
> > This expression contains a magic number. Consider defining it to a well named constant. [MagicNumber]
> > This is thrown at a line where I am assigning a value to a variable, I am unable to understand how to fix this.
> 
> @shankarpriyank You can suppress this by annotating method supress warnings or make a constant value.

Thanks, I understand the supressing alternative, but what do you mean by >make a constant value? 

Also can you please help me with the failing tests, I don't have much experience in testing,but after having a look at the log it seems to me like that the mock objects in some ways different to the real one.

gouri-panda commented:
>what do you mean by > make a constant value

Constant value means create a constant variable. The error is simply means you can't put only numbers in arguments. By doing that we can put different number in different locations by mistake. In order  to not make this mistake we can put a constant variable and use it everywhere else.

shankarpriyank commented:
I guess I have done that same thing
```
private val viewSpacingLeftForLink = 0
  private val viewSpacingRightForLink = 0
  private val viewSpacingTopForLink = 10
  private val viewSpacingBottomForLink = 0
```
I declared these variables and then passed them as arguments in this function

```
         setView(
            textView,
            viewSpacingLeftForLink,
            viewSpacingTopForLink,
            viewSpacingRightForLink,
            viewSpacingBottomForLink
          )
```

shankarpriyank commented on core/src/main/java/org/kiwix/kiwixmobile/core/utils/dialog/AlertDialogShower.kt:
```diff
@@ -20,20 +20,34 @@ package org.kiwix.kiwixmobile.core.utils.dialog
-  override fun show(dialog: KiwixDialog, vararg clickListeners: () -> Unit) {
-    create(dialog, *clickListeners).show()
-  }
+  private val viewSpacingLeftForLink = 0
```
@gouri-panda Please have a look here

gouri-panda replied:
Ok. Maybe we can do better at naming variable  :)

shankarpriyank replied:
Yeah, sure any suggestions regarding the names?

gouri-panda replied:
Anything that's more easily readable are welcome!

MohitMaliFtechiz replied:
@gouri-panda ,
I have rename these variables.

gouri-panda commented:
@kelson42 Are you ok with the UI?

kelson42 commented:
@gouri-panda AFAIK the UI has not been fixed, so would like to see a proper UI, the current one buttons are not properly aligned.

gouri-panda commented:
@kelson42 Ok. @shankarpriyank Please change the UI as @kelson42 mentioned. Then I'll review the other parts :) Also, the other errors haven't been fixed yet.

kelson42 commented:
If the alignment problen can not be fixed, please open a dedicated ticket. Then we could finally review and merge this PR.

gouri-panda commented:
@shankarpriyank This test is failing. 
>ExternalLinkOpenerTest > alertDialogShower does not open link if negative-button is clicked$core_debug() FAILED
    io.mockk.MockKException at ExternalLinkOpenerTest.kt:68
    
The `ExternalLinkOpener` has changed. If the change is intentional then you can modify these tests. If you want more tests that you have changed you can add more :slightly_smiling_face:

shankarpriyank commented:
Hey @kelson42 I am still working on modifying the tests so that they pass. I donot have much experience in testing and this is the first time I am encountering mocks so its a bit challenging. I could use some help to fix the tests.

kelson42 commented:
@shankarpriyank Sorry, my bad, I had the feeling you were in a dead-end.

4shutosh commented on core/src/main/java/org/kiwix/kiwixmobile/core/utils/dialog/AlertDialogShower.kt:
```diff
@@ -51,6 +66,33 @@ class AlertDialogShower @Inject constructor(private val activity: Activity) :
+            clipboard?.setPrimaryClip(clip)
+            Toast.makeText(
+              activity.baseContext,
+              "External Link Copied to Clipboard",
```
hardcoded strings should be extracted I think

shankarpriyank replied:
Oh, yes will fix it!

MohitMaliFtechiz replied:
@4shutosh ,
done

4shutosh commented on core/src/main/java/org/kiwix/kiwixmobile/core/utils/dialog/AlertDialogShower.kt:
```diff
@@ -51,6 +66,33 @@ class AlertDialogShower @Inject constructor(private val activity: Activity) :
+            ).show()
+            true
+          }
+          textView.text = Html.fromHtml("</br><a href=$url> <b>$url</b>")
```
Suggestion: Instead of converting a link to html text, you can always use spannable string, and implement a custom click behaviour here, instead of leaving it to OS.

kelson42 commented:
What is the status here?

shankarpriyank commented:
The functionality has been implemented, but I am unable to modify the tests so that they pass. Took some help from @4shutosh but still no success.

kelson42 commented:
@MohitMaliDeveloper Any chance you can complete and review his PR?

MohitMaliFtechiz commented:
> @MohitMaliDeveloper Any chance you can complete and review his PR?

ok

kelson42 commented:
@MohitMaliFtechiz Have you finished the necessary work? Shoukd you remove the draft mode?

MohitMaliFtechiz commented:
> @MohitMaliFtechiz Have you finished the necessary work? Shoukd you remove the draft mode?

@kelson42 ,
I have fixed test and lint failures.

> AFAIK the UI has not been fixed, so would like to see a proper UI, the current one buttons are not properly aligned.

Ui changes are left.

kelson42 commented:
@MohitMaliFtechiz OK, who will do that?

MohitMaliFtechiz commented:
> @MohitMaliFtechiz OK, who will do that?

I'm working on it.

MohitMaliFtechiz commented:
@kelson42 ,
I'm testing the UI. I found this is a ```AlertDialog``` and it's button automatically changed according to every screen size and how many space left in screen. 

Example :- 

**In Portrait Mode**

![Screenshot_2023-01-28-18-35-02-293_org kiwix kiwixmobile 1](https://user-images.githubusercontent.com/34593983/215268013-c6ac8906-e175-4630-be56-afc59088fb2c.jpg)

**In Landscape Mode**

![Screenshot_2023-01-28-18-35-12-701_org kiwix kiwixmobile 1](https://user-images.githubusercontent.com/34593983/215268044-f3c6d098-b855-4fc0-aca4-456760d8fe9f.jpg)

If we want to change the UI then we need to use a custom layout with ```Custom Dialog``` instead of Android ```AlertDialog```.

kelson42 commented:
@MohitMaliDeveloper Not too bad, can we reduce the big white space? Please rebase the PR as well.

MohitMaliFtechiz commented:
> @MohitMaliDeveloper Not too bad, can we reduce the big white space? 

@kelson42 , unfortunately not, it's automatically takes height according to it's views.

> Please rebase the PR as well.
   
I have re base the PR and resolved the conflicts.

kelson42 commented:
@MohitMaliDeveloper Is that ready to review (please rebase and fix conflicts)?

MohitMaliFtechiz commented:
> @MohitMaliDeveloper Is that ready to review (please rebase and fix conflicts)?

@kelson42 , i'm waiting for your feedback on above comment. that's why this PR in draft, if that is okay for you then this PR is ready for review.

kelson42 commented:
@MohitMaliFtechiz LGTM. If you request something from me, please specificaly ask with a question mark.

MohitMaliFtechiz commented:
@kelson42 , I have resolved the conflicts. @gouri-panda this PR is ready for review.

gouri-panda commented:
@MohitMaliDeveloper Thanks! Will do!

gouri-panda commented on core/src/main/java/org/kiwix/kiwixmobile/core/utils/dialog/AlertDialogShower.kt:
```diff
@@ -51,6 +69,33 @@ class AlertDialogShower @Inject constructor(private val activity: Activity) :
+          }
+          textView.text = Html.fromHtml("</br><a href=$url> <b>$url</b>")
+
+          setView(
+            textView,
+            externalLinkLeftMargin,
+            externalLinkTopMargin,
+            externalLinkRightMargin,
+            externalLinkBottomMargin
+          )
+        }
```
This method is deprecated. We can use the alternative one. Also, we can get rid of this restricted API warning.

MohitMaliFtechiz replied:
@gouri-panda , i have removed this deprecated method and restricted API warning.

gouri-panda reviewed (commented):
@MohitMaliDeveloper  @shankarpriyank I overviewed the PR.  I leave some comments about this. The external link colour doesn't look good on the dark theme. Just a small friendly reminder, please don't change the unrelated code to tickets as you did here, changing the licence and renaming the function names.

MohitMaliFtechiz commented:
@gouri-panda ,

> changing the licence

* It's automatically changed when we move file to another package. I'm reverting the ```Copyright (c)``` to ```2020```.

> renaming the function names

* Now ```ExternalLinkOpenerTest```  file is in ```androidTest``` package, here these function name is not allowed that's why i had changed it.

![Screenshot from 2023-02-03 11-51-02](https://user-images.githubusercontent.com/34593983/216527709-d7dc6cd0-047e-468d-9615-55024b21f89b.png)

gouri-panda commented:
@MohitMaliDeveloper Did you push it? I don't see recent changes.

MohitMaliFtechiz commented:
> @MohitMaliDeveloper Did you push it? I don't see recent changes.

hi @gouri-panda ,
I have just push the all latest changes(as you requested).

gouri-panda reviewed (approved):
@shankarpriyank @MohitMaliDeveloper Thanks for the work 🎉 !
